### DISCUSSION TITLE
Motorbike physics

### DISCUSSION THREAD

**Discussion by Ono-Sendai** [jrouwe/JoltPhysics#302] [General]
Hi Jorrit, all,
So far I have got car physics working pretty decently in my engine (see https://twitter.com/SubstrataVr/status/1584551939651477504 )

I'm quite keen to get motorbike physics in there as well.   What would be the best approach for this in Jolt?
As I understand it, true motorbike physics is quite complicated, involving countersteering when entering corners, moving the rider's centre of mass, curved tire profiles etc.   So I think usually a much simplified physics model is used.
Maybe we just use some kind of artificial force or constraint to keep the bike from rolling to the side too much?  (e.g. to keep it from falling over sideways).  And when the bike goes around a corner, we artificially tilt the bike over?

Does the current Jolt vehicle code handle 2-wheeled vehicles?  Or are there some assumptions about axles or related built in?

**Comment by jrouwe** [comment]
Hello,

I've seen your tweet, pretty cool! 

I don't have any motorcycle physics experience, so I'm not sure what the best way is. In any case you can make a wheeled vehicle with only 1 front and 1 back wheel. For your differential set mLeftWheel = back wheel index and leave mRightWheel = -1, then set mLeftRightSplit = 0 to apply all torque to the left wheel. Try setting a limiting angle on VehicleConstraintSettings::mMaxPitchRollAngle to prevent the bike from toppling over (note that this is a single angle right now, for a bike you probably want no/a large limit for pitch and only a limit for roll, I haven't made this distinction and you can't turn it off in case of a crash either at the moment).

Jorrit.

**Comment by Ono-Sendai** [comment]
Hi Jorrit, 
looking at this again (see https://twitter.com/SubstrataVr/status/1631501918068707328)
I think the obvious approach with the angle constraints is to split them into individual pitch and roll constraints as you mentioned.
Will you accept a patch doing that?
I could also make a new constraint that wraps VehicleConstraint, but that seems more complicated.
Cheers,
  Nick

**Reply by jrouwe** [reply]
If you mean splitting up ```AngleConstraintPart mPitchRollPart``` in a ```AngleConstraintPart mPitchPart``` and ```AngleConstraintPart mRollPart`` then yes I would be willing to accept that patch.

**Reply by jrouwe** [reply]
Video looks nice btw! It feels a bit unnatural the way the bike is leaning over all the time. Since the wheels have no angular momentum, there's nothing that is going to keep the bike upright. One potential thing you could do is actually have 2 front and 2 back wheels that are very close (and obviously only render 1 front and 1 back wheel) so that the bike will stay upright more. Alternatively, a 3rd AngleConstraintPart could be added that serves as a motor to drive the bike to a vertical position (just like HingeConstraint::mMotorConstraintPart). The strength of the motor could vary with velocity so that when you stop, you lean over and when you accelerate the bike becomes more vertical. (Just brainstorming here b.t.w.)

**Reply by Ono-Sendai** [reply]
> If you mean splitting up `AngleConstraintPart mPitchRollPart` in a `AngleConstraintPart mPitchPart` and ```AngleConstraintPart mRollPart`` then yes I would be willing to accept that patch.

Yeah that's what I mean.

**Reply by Ono-Sendai** [reply]
> Video looks nice btw! It feels a bit unnatural the way the bike is leaning over all the time. Since the wheels have no angular momentum, there's nothing that is going to keep the bike upright. One potential thing you could do is actually have 2 front and 2 back wheels that are very close (and obviously only render 1 front and 1 back wheel) so that the bike will stay upright more. Alternatively, a 3rd AngleConstraintPart could be added that serves as a motor to drive the bike to a vertical position (just like HingeConstraint::mMotorConstraintPart). The strength of the motor could vary with velocity so that when you stop, you lean over and when you accelerate the bike becomes more vertical. (Just brainstorming here b.t.w.)

Yeah I think I will try some kind of torque to rotate to the desired orientation, which will depend on which way the user is steering.
